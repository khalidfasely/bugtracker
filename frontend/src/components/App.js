function App() {
  return (
    <div className="App">
      <header>
        Hello world from React        
      </header>
    </div>
  );
}

export default App;
