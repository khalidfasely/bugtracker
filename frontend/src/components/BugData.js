const BugData = ({ bug }) => (
    <div>
        <p>{bug?.id}</p>
        <p>{bug?.user?.username}</p>
        <p>{bug?.title}</p>
        <p>{bug?.description}</p>
        <p>{bug?.time}</p>
        <p>{bug?.classification}</p>
        <p>{`${bug?.active}`}</p>
        <div>
            {bug?.admins?.map(admin => <span key={admin} >{admin} </span>)}
        </div>
        <br />
        <div>
            {bug?.users_with?.map(user => <span key={user} >{user} </span>)}
        </div>
    </div>
);

export default BugData;