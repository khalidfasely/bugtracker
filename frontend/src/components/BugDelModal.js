import Modal from "react-modal";

const BugDelModal = ({ delModalOpen, setDelModalOpen, deleteBug }) => (
    <Modal
        isOpen={delModalOpen}
        onRequestClose={() => setDelModalOpen(false)}
    >
        If you delete this bug there is no way to return it!
        <button onClick={deleteBug} data-testid='delete_button'>Delete</button>
    </Modal>
);

export default BugDelModal;