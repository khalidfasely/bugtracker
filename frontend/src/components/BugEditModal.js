import Modal from "react-modal";

import NewBug from "./NewBug";

const BugEditModal = ({
    editModalOpen, setEditModalOpen,
    users, bug
}) => (
    <Modal
        isOpen={editModalOpen}
        onRequestClose={() => setEditModalOpen(false)}
    >
        <NewBug users={users} isEdit={true} bug={bug} setEditModalOpen={setEditModalOpen} />
    </Modal>
);

export default BugEditModal;