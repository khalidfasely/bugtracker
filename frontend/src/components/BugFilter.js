import { connect } from "react-redux";
import { setTextFilter, sortByDefault, sortByPriority } from "../actions/bugFilter";

export const BugFilter = ({ filters, setTextFilter, sortByDefault, sortByPriority }) => {
    const onSortByChange = (e) => {
        if (e.target.value === 'priority') {
            sortByPriority();
            return;
        };
        sortByDefault();
    };

    return (
        <div className="bugs-filter">
            <div className="input-group__item">
                <input
                    data-testid='search_input'
                    type="text"
                    className="text-input"
                    placeholder="Search bugs"
                    value={filters.text}
                    maxLength={35}
                    onChange={(e) => setTextFilter(e.target.value)}
                />
            </div>
            <div className="input-group__item">
                <select
                    data-testid='select_input'
                    className="select"
                    value={filters.sortBy}
                    onChange={onSortByChange}
                >
                    <option value="">Date</option>
                    <option value="priority">Priority</option>
                </select>
            </div>
        </div>
    );
};

const mapStateToProps = (state) => ({
    filters: state.bugFilter
});

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByPriority: () => dispatch(sortByPriority()),
    sortByDefault: () => dispatch(sortByDefault())
});

export default connect(mapStateToProps, mapDispatchToProps)(BugFilter);