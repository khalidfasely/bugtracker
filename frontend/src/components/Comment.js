import { useState } from "react";
import { connect } from "react-redux";

import NewComment from "./NewComment";
import { startSetDeleteComment } from "../actions/bug";
import CommentDelModal from "./CommentDelModal";

export const Comment = ({ comment, uname, startSetDeleteComment }) => {
    const [isEdit, setIsEdit] = useState(false);
    const [delModalOpen, setDelModalOpen] = useState(false);

    const deleteComment = () => {
      startSetDeleteComment(comment.id).then(() => {
        setDelModalOpen(false);
      });
    };
    
    if (isEdit) {
      return <NewComment commentEdit={comment.content} commentId={comment.id} setIsEdit={setIsEdit} />
    };
    
    return (
        <div>
          <p>{comment.id}</p>
          <p>{comment.content}</p>
          <p>{comment.user.username}</p>
          <p>{comment.time}</p>
          {
            uname === comment.user.username &&
            <div>
              <button onClick={() => setIsEdit(true)} data-testid='edit_button' >Edit</button>
              <button onClick={() => setDelModalOpen(true)}>Delete</button>
            </div>
          }
          <CommentDelModal
            delModalOpen={delModalOpen}
            setDelModalOpen={setDelModalOpen}
            deleteComment={deleteComment}
          />
        </div>
    );
};

const mapStateToProps = (state) => ({
  uname: state.auth.uname
});

const mapDispatchToProps = (dispatch) => ({
  startSetDeleteComment: (cid) => dispatch(startSetDeleteComment(cid))
});

export default connect(mapStateToProps, mapDispatchToProps)(Comment);