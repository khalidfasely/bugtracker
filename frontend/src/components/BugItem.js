import { useState, useEffect } from "react";
import { Link, useNavigate } from "react-router-dom";
import { connect } from "react-redux";

import Comment from "./Comment";
import BugData from "./BugData";
import NewComment from "./NewComment";
import { startSetDeleteBug } from "../actions/project";
import { startSetSelectUsers } from '../actions/selectUsers';
import BugDelModal from "./BugDelModal";
import BugEditModal from "./BugEditModal";

export const BugItem = ({
        uname, bug, comments, startSetDeleteBug,
        users, startSetSelectUsers
    }) => {
    const history = useNavigate();

    const [delModalOpen, setDelModalOpen] = useState(false);
    const [editModalOpen, setEditModalOpen] = useState(false);

    // Set select users
    useEffect(() => startSetSelectUsers(bug.on_project), []);

    const deleteBug = () => {
        startSetDeleteBug(bug.id).then(() => {
            history(`/project/${bug.on_project}`);
            setDelModalOpen(false);
        });
    };

    return (
        <>
            {
                bug?.users_with?.includes(uname) ?
                <div>
                    <BugData bug={bug} />
                    <hr />
                    {
                        uname===bug.user.username &&
                        <div>
                            <button onClick={() => setDelModalOpen(true)}>Delete</button>
                            <button onClick={() => setEditModalOpen(true)}>Edit</button>
                        </div>
                    }
                    <hr />
                    <NewComment bugId={bug.id} />
                    <hr />
                    {comments?.map(comment => <div key={comment.id}><Comment comment={comment} /></div>)}
                </div> :
                <div>Not able to see this bug! <Link to={`/project/${bug.on_project}`}>Create one!</Link></div>
            }
            <BugDelModal
                delModalOpen={delModalOpen}
                setDelModalOpen={setDelModalOpen}
                deleteBug={deleteBug}
            />
            <BugEditModal
                editModalOpen={editModalOpen}
                setEditModalOpen={setEditModalOpen}
                users={users}
                bug={bug}
            />
        </>
    );
}

const mapStateToProps = (state) => ({
    uname: state.auth.uname,
    bug: state.bugItem.bug,
    comments: state.bugItem.comments,
    users: state.selectUsers
});

const mapDispatchToProps = (dispatch) => ({
    startSetDeleteBug: (bid) => dispatch(startSetDeleteBug(bid)),
    startSetSelectUsers: (pid) => dispatch(startSetSelectUsers(pid))
});

export default connect(mapStateToProps, mapDispatchToProps)(BugItem);