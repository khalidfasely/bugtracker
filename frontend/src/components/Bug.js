import { useState, useEffect } from "react";
import { connect } from "react-redux";
import { useParams } from "react-router-dom";
import { startSetBug } from "../actions/bug";

import BugItem from './BugItem';

export const Bug = ({ startSetBug }) => {
    const params = useParams();

    const [ableToSee, setAbleToSee] = useState(false);
    const [notFound, setNotFound] = useState('');

    useEffect(() => {
        startSetBug(params.pid, params.bid)
        .then((result) => {
            setNotFound('');

            if (result.message) {
                setNotFound(result.message);
            };
        })
        .then(() => setAbleToSee(true))
        .catch(er => console.error(er));
    }, [params])


    if (notFound) {
        return <div>Bug Not Found! -- {notFound}</div>;
    };

    return (
        <div>
            {
                ableToSee ?
                <div><BugItem /></div> :
                <div>You're not able to see this bug</div>
            }
        </div>
    );
};

const mapDispatchToProps = (dispatch) => ({
    startSetBug: (pid, bid) => dispatch(startSetBug(pid, bid))
});

export default connect(undefined, mapDispatchToProps)(Bug);