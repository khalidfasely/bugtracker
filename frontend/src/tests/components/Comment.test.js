import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { Comment } from '../../components/Comment';
import comments from "../fixtures/comments";

jest.mock('../../components/CommentDelModal', () => () => <div>CommentDelModal</div>);
jest.mock('../../components/NewComment', () => () => <div>NewComment</div>);

let props;

beforeEach(() => {
    props = {
        comment: comments[0],
        uname: '1',
        startSetDeleteComment: jest.fn(() => Promise.resolve())
    }
});

test('Should render Comment component correctly with user not the comment creator', () => {
    const { asFragment } = render(<Comment {...{...props, uname: 'Anonymous'}} />);
    expect(asFragment()).toMatchSnapshot();
});

test('Should render Comment component correctly with user is the comment creator', () => {
    const { asFragment } = render(<Comment {...props} />);
    expect(asFragment()).toMatchSnapshot();
});

test('Should edit button fire correctly', () => {
    const { asFragment, getByTestId } = render(<Comment {...props} />);
    const buttonEditEl = getByTestId('edit_button');

    fireEvent.click(buttonEditEl);

    expect(asFragment()).toMatchSnapshot();
});