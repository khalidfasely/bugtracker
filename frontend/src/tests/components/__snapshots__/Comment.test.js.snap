// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should edit button fire correctly 1`] = `
<DocumentFragment>
  <div>
    NewComment
  </div>
</DocumentFragment>
`;

exports[`Should render Comment component correctly with user is the comment creator 1`] = `
<DocumentFragment>
  <div>
    <p>
      1
    </p>
    <p>
      Comment
    </p>
    <p>
      1
    </p>
    <p>
      now
    </p>
    <div>
      <button
        data-testid="edit_button"
      >
        Edit
      </button>
      <button>
        Delete
      </button>
    </div>
    <div>
      CommentDelModal
    </div>
  </div>
</DocumentFragment>
`;

exports[`Should render Comment component correctly with user not the comment creator 1`] = `
<DocumentFragment>
  <div>
    <p>
      1
    </p>
    <p>
      Comment
    </p>
    <p>
      1
    </p>
    <p>
      now
    </p>
    <div>
      CommentDelModal
    </div>
  </div>
</DocumentFragment>
`;
